README.md
# flutter_mjpeg

Flutter widget to show mjpeg stream from URL

This is a full dart implementation of MJPEG reader. No native involve.

## Usage

```
Mjpeg(
  stream: 'http://192.168.1.24:8080/video.cgi',
)
```

## API
The [`Mjpeg` widget](https://pub.dev/documentation/flutter_mjpeg/latest/flutter_mjpeg/Mjpeg-class.html) expects a `stream` parameter with the HTTP URL of the MJPEG stream and can handle the following additional parameters:

Parameter | Description
--- | ---
`isLive` | Whether or not the stream should be loaded continuously
`timeout` | HTTP Timeout when fetching the MJPEG stream
`width` | Force width
`height` | Force height
`error` | Error builder used when an error occurred
`loading` | Loading builder used until first frame arrived
`fit` | The `boxFit` of the image
`headers` | A map of headers to send in the HTTP request
`httpClient` | Used to give a custom httpClient, for example `DigestAuthClient()` from [http_auth](https://pub.dev/packages/http_auth). Defaults to `Client()` from [http](https://pub.dev/packages/http).
`preprocessor` | Used to apply preprocessing to each frame of the MJPEG stream before it is sent to [Image](https://api.flutter.dev/flutter/widgets/Image-class.html) for rendering. Defaults to `MjpegPreprocessor()`, which passes each frame without modification. 


CHANGELOG.md
## [2.0.4] - 15/07/2023

* prevent errors when widget not mounted

## [2.0.3] - 04/10/2022

* add preprocessor field to process frame if needed

## [2.0.2] - 26/09/2022

* fix image decoding error
* better doc

## [2.0.1] - 11/06/2021

* fix glitch on some streams
* add stack param on onError callback

## [2.0.0] - 21/03/2021

* migrate to null safety
* expose timeout field

## [1.3.0] - 22/01/2021

* improve framerate
* update deps

## [1.2.9] - 10/08/2020

* correct deps version

## [1.2.8] - 29/06/2020

* correctly handle http errors (other than >= 200 > 300)

## [1.2.7] - 08/06/2020

* fix crash when widget was disposed.

## [1.2.6] - 26/03/2020

* add timeout on open connexion to prevent forever loading state

## [1.2.5] - 03/10/2019

* improve pub scoring 

## [1.2.4] - 02/10/2019

* close http client when widget not on screen

## [1.2.3] - 02/10/2019

* close http client when dispose 

## [1.2.2] - 20/09/2019

* fix some errors happening in some mjpeg stream 

## [1.2.1] - 18/09/2019

* add possibility to pass headers to the stream request

## [1.2.0] - 16/09/2019

* add compatibility with flutter web by using http package

## [1.1.1] - 10/07/2019

* update flutter_hook to be compatible with last stable

## [1.1.0] - 18/06/2019

* add a play/pause state

## [1.0.1] - 16/06/2019

* apply width and height on all states

## [1.0.0] - 16/06/2019

* initial release.


test/flutter_mjpeg_test.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';

const _trigger = 0xFF;
const _soi = 0xD8;
const _eoi = 0xD9;

void main() {
  // ignore_for_file: close_sinks
  final streamController = StreamController<List<int>>.broadcast();

  streamController.stream.transform(JpegSplitter()).listen((data) {
    print(data[0]);
    print(data[1]);
    print(data[data.length - 1]);
    print(data[data.length - 2]);
  });

  //streamController.add([_trigger, _soi, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, _trigger, _eoi]);
  streamController.add([23, 23, _trigger, _soi, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, 23, _trigger]);
  streamController.add([_eoi]);

  test('adds one to input values', () {});
}

/// A [StreamTransformer] that splits a [List<int>] into individual mjpeg frame.
///
/// A line is terminated by either a CR (U+000D), a LF (U+000A), a
/// CR+LF sequence (DOS line ending),
/// and a final non-empty line can be ended by the end of the string.
///
/// The returned lines do not contain the line terminators.
class JpegSplitter extends StreamTransformerBase<List<int>, List<int>> {
  const JpegSplitter();

  ChunkedConversionSink<List<int>> startChunkedConversion(Sink<List<int>> sink) {
    return _JpegSplitterSink(sink as ChunkedConversionSink<List<int>?>);
  }

  Stream<List<int>> bind(Stream<List<int>> stream) {
    return Stream<List<int>>.eventTransformed(stream, (EventSink<List<int>> sink) {
      return _JpegSplitterEventSink(sink);
    });
  }
}

class _JpegSplitterSink extends ChunkedConversionSink<List<int>> {
  final ChunkedConversionSink<List<int>?> _sink;

  /// The carry-over from the previous chunk.
  ///
  /// If the previous slice ended in a line without a line terminator,
  /// then the next slice may continue the line.
  List<int>? _carry = [];

  _JpegSplitterSink(this._sink);

  void close() {
    print('close');
    if (_carry != null) {
      _sink.add(_carry);
      _carry = null;
    }
    _sink.close();
  }

  @override
  void add(List<int> chunk) {
    if (_carry!.isNotEmpty && _carry!.last == _trigger) {
      if (chunk.first == _eoi) {
        _carry!.add(chunk.first);
        _sink.add(_carry);
        _carry = [];
      }
    }

    for (var i = 0; i < chunk.length - 1; i++) {
      final d = chunk[i];
      final d1 = chunk[i + 1];

      if (d == _trigger && d1 == _soi) {
        _carry!.add(d);
      } else if (d == _trigger && d1 == _eoi) {
        _carry!.add(d);
        _carry!.add(d1);
        _sink.add(_carry);
        _carry = [];
      } else if (_carry!.isNotEmpty) {
        _carry!.add(d);
        if (i == chunk.length - 2) {
          _carry!.add(d1);
        }
      }
    }
  }
}

class _JpegSplitterEventSink extends _JpegSplitterSink implements EventSink<List<int>> {
  final EventSink<List<int>> _eventSink;

  _JpegSplitterEventSink(EventSink<List<int>> eventSink)
      : _eventSink = eventSink,
        super(ChunkedConversionSink<List<int>>.withCallback((accumulated) {
          print('accumulated');
        }));

  void addError(Object o, [StackTrace? stackTrace]) {
    _eventSink.addError(o, stackTrace);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_hooks/flutter_hooks.dart';
import 'package:flutter_mjpeg/flutter_mjpeg.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends HookWidget {
  @override
  Widget build(BuildContext context) {
    final isRunning = useState(true);
    return Scaffold(
      appBar: AppBar(
        title: Text('Flutter Demo Home Page'),
      ),
      body: Column(
        children: <Widget>[
          Expanded(
            child: Center(
              child: Mjpeg(
                isLive: isRunning.value,
                error: (context, error, stack) {
                  print(error);
                  print(stack);
                  return Text(error.toString(), style: TextStyle(color: Colors.red));
                },
                stream:
                'http://uk.jokkmokk.jp/photo/nr4/latest.jpg', //'http://192.168.1.37:8081',
              ),
            ),
          ),
          Row(
            children: <Widget>[
              ElevatedButton(
                onPressed: () {
                  isRunning.value = !isRunning.value;
                },
                child: Text('Toggle'),
              ),
              ElevatedButton(
                onPressed: () {
                  Navigator.of(context).push(MaterialPageRoute(
                      builder: (context) => Scaffold(
                            appBar: AppBar(),
                          )));
                },
                child: Text('Push new route'),
              ),
            ],
          ),
        ],
      ),
    );
  }
}


lib/src/mjpeg.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:flutter_hooks/flutter_hooks.dart';
import 'package:http/http.dart';
import 'package:visibility_detector/visibility_detector.dart';

class _MjpegStateNotifier extends ChangeNotifier {
  bool _mounted = true;
  bool _visible = true;

  _MjpegStateNotifier() : super();

  bool get mounted => _mounted;

  bool get visible => _visible;

  set visible(value) {
    _visible = value;
    notifyListeners();
  }

  @override
  void dispose() {
    _mounted = false;
    notifyListeners();
    super.dispose();
  }
}

/// A preprocessor for each JPEG frame from an MJPEG stream.
class MjpegPreprocessor {
  List<int>? process(List<int> frame) => frame;
}

/// An Mjpeg.
class Mjpeg extends HookWidget {
  final String stream;
  final BoxFit? fit;
  final double? width;
  final double? height;
  final bool isLive;
  final Duration timeout;
  final WidgetBuilder? loading;
  final Client? httpClient;
  final Widget Function(BuildContext contet, dynamic error, dynamic stack)?
      error;
  final Map<String, String> headers;
  final MjpegPreprocessor? preprocessor;

  const Mjpeg({
    this.httpClient,
    this.isLive = false,
    this.width,
    this.timeout = const Duration(seconds: 5),
    this.height,
    this.fit,
    required this.stream,
    this.error,
    this.loading,
    this.headers = const {},
    this.preprocessor,
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final image = useState<MemoryImage?>(null);
    final state = useMemoized(() => _MjpegStateNotifier());
    final visible = useListenable(state);
    final errorState = useState<List<dynamic>?>(null);
    final isMounted = useIsMounted();
    final manager = useMemoized(
        () => _StreamManager(
              stream,
              isLive && visible.visible,
              headers,
              timeout,
              httpClient ?? Client(),
              preprocessor ?? MjpegPreprocessor(),
              isMounted,
            ),
        [
          stream,
          isLive,
          visible.visible,
          timeout,
          httpClient,
          preprocessor,
          isMounted
        ]);
    final key = useMemoized(() => UniqueKey(), [manager]);

    useEffect(() {
      errorState.value = null;
      manager.updateStream(context, image, errorState);
      return manager.dispose;
    }, [manager]);

    if (errorState.value != null) {
      return SizedBox(
        width: width,
        height: height,
        child: error == null
            ? Center(
                child: Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: Text(
                    '${errorState.value}',
                    textAlign: TextAlign.center,
                    style: TextStyle(color: Colors.red),
                  ),
                ),
              )
            : error!(context, errorState.value!.first, errorState.value!.last),
      );
    }

    if (image.value == null) {
      return SizedBox(
          width: width,
          height: height,
          child: loading == null
              ? Center(child: CircularProgressIndicator())
              : loading!(context));
    }

    return VisibilityDetector(
      key: key,
      child: Image(
        image: image.value!,
        width: width,
        height: height,
        gaplessPlayback: true,
        fit: fit,
      ),
      onVisibilityChanged: (VisibilityInfo info) {
        if (visible.mounted) {
          visible.visible = info.visibleFraction != 0;
        }
      },
    );
  }
}

class _StreamManager {
  static const _trigger = 0xFF;
  static const _soi = 0xD8;
  static const _eoi = 0xD9;

  final String stream;
  final bool isLive;
  final Duration _timeout;
  final Map<String, String> headers;
  final Client _httpClient;
  final MjpegPreprocessor _preprocessor;
  final bool Function() _mounted;
  // ignore: cancel_subscriptions
  StreamSubscription? _subscription;

  _StreamManager(this.stream, this.isLive, this.headers, this._timeout,
      this._httpClient, this._preprocessor, this._mounted);

  Future<void> dispose() async {
    if (_subscription != null) {
      await _subscription!.cancel();
      _subscription = null;
    }
    _httpClient.close();
  }

  void _sendImage(BuildContext context, ValueNotifier<MemoryImage?> image,
      ValueNotifier<dynamic> errorState, List<int> chunks) async {
    // pass image through preprocessor sending to [Image] for rendering
    final List<int>? imageData = _preprocessor.process(chunks);
    if (imageData == null) return;

    final imageMemory = MemoryImage(Uint8List.fromList(imageData));
    if (_mounted()) {
      errorState.value = null;
      image.value = imageMemory;
    }
  }

  void updateStream(BuildContext context, ValueNotifier<MemoryImage?> image,
      ValueNotifier<List<dynamic>?> errorState) async {
    try {
      final request = Request("GET", Uri.parse(stream));
      request.headers.addAll(headers);
      final response = await _httpClient.send(request).timeout(
          _timeout); //timeout is to prevent process to hang forever in some case

      if (response.statusCode >= 200 && response.statusCode < 300) {
        var _carry = <int>[];
        _subscription = response.stream.listen((chunk) async {
          if (_carry.isNotEmpty && _carry.last == _trigger) {
            if (chunk.first == _eoi) {
              _carry.add(chunk.first);
              _sendImage(context, image, errorState, _carry);
              _carry = [];
              if (!isLive) {
                dispose();
              }
            }
          }

          for (var i = 0; i < chunk.length - 1; i++) {
            final d = chunk[i];
            final d1 = chunk[i + 1];

            if (d == _trigger && d1 == _soi) {
              _carry = [];
              _carry.add(d);
            } else if (d == _trigger && d1 == _eoi && _carry.isNotEmpty) {
              _carry.add(d);
              _carry.add(d1);

              _sendImage(context, image, errorState, _carry);
              _carry = [];
              if (!isLive) {
                dispose();
              }
            } else if (_carry.isNotEmpty) {
              _carry.add(d);
              if (i == chunk.length - 2) {
                _carry.add(d1);
              }
            }
          }
        }, onError: (error, stack) {
          try {
            if (_mounted()) {
              errorState.value = [error, stack];
              image.value = null;
            }
          } catch (ex) {}
          dispose();
        }, cancelOnError: true);
      } else {
        if (_mounted()) {
          errorState.value = [
            HttpException('Stream returned ${response.statusCode} status'),
            StackTrace.current
          ];
          image.value = null;
        }
        dispose();
      }
    } catch (error, stack) {
      // we ignore those errors in case play/pause is triggers
      if (!error
          .toString()
          .contains('Connection closed before full header was received')) {
        if (_mounted()) {
          errorState.value = [error, stack];
          image.value = null;
        }
      }
    }
  }
}
